### PR TITLE
PPML FGBoost (GBDT)

### PR DESCRIPTION
**Pull request by Litchilitchy**
* FGBoost code and example
* some refactor of previous FGBoost proto
* some rename of NN example

This PR is mainly about API generalize. Not many lower-level refactors in this PR. And the example still has something to improve.

### CODE REVIEW

**Comment by jason-dai** [comment]
@qiuxin2012 please take a look

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/Split.scala [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.ppml.vfl.fgboost
+
+import com.intel.analytics.bigdl.ppml.generated.FGBoostServiceProto.DataSplit
+import org.apache.log4j.Logger
```
Comment (qiuxin2012) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/Split.scala:20]: log4j

File: scala/ppml/demo/data/house-prices.csv [hunk 2]
```diff
@@ -0,0 +1,1461 @@
+Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,OverallQual,OverallCond,YearBuilt,YearRemodAdd,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,MasVnrArea,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinSF1,BsmtFinType2,BsmtFinSF2,BsmtUnfSF,TotalBsmtSF,Heating,HeatingQC,CentralAir,Electrical,1stFlrSF,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,KitchenQual,TotRmsAbvGrd,Functional,Fireplaces,FireplaceQu,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond,PavedDrive,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
```
Comment (qiyuangong) [inline scala/ppml/demo/data/house-prices.csv]: Not a good idea to add dataset into github repo

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/FLClient.scala [hunk 3]
```diff
@@ -36,25 +36,27 @@ class FLClient(val _args: Array[String]) extends GrpcClientBase(_args) {
   protected var taskID: String = "taskID"
   var psiStub: PSIStub = null
   var nnStub: NNStub = null
-
+  var fgbostStub: FGBoostStub = null
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/FLClient.scala:39]: fgbStub

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      var st = System.currentTimeMillis()
+      val bestLocalSplit = tree.findBestSplit()
+      logger.debug(s"Find best split cost ${(System.currentTimeMillis() - st) / 1000f} s")
+      st = System.currentTimeMillis()
+      val bestSplit = getBestSplitFromServer(bestLocalSplit)
+      logger.debug(s"Sync split cost ${(System.currentTimeMillis() - st) / 1000f} s")
+      st = System.currentTimeMillis()
+      val isLocalSplit = bestLocalSplit.getClientID == bestSplit.getClientID
+      // If this split is in local dataset
+      val updateCondition = if (continuous) {
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala:153]: stopCondition?

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      } else  bestSplit.featureID == -1
+      if (updateCondition) {
+        logger.warn("Fail to split on current node")
+        logger.info(s"Set Leaf gain = ${bestSplit.gain}")
+        // Add current node to leaf
+        tree.setLeaf(tree.nodes(bestSplit.nodeID))
+        logger.debug(s"Set leaf ${(System.currentTimeMillis() - st) / 1000f} s")
+        st = System.currentTimeMillis()
+      } else {
+        // Then, apply split & upload to VFL Server
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala:164]: "upload to VFL Server"?

File: scala/ppml/src/main/java/com/intel/analytics/bigdl/ppml/vfl/FGBoostStub.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    public DownloadResponse downloadTable(String modelName, int flVersion) {
+        logger.info("Download the following data:");
+        TableMetaData metadata = TableMetaData.newBuilder()
+                .setName(modelName).setVersion(flVersion + 1).build();
+        DownloadTableRequest downloadRequest = DownloadTableRequest.newBuilder().setMetaData(metadata).build();
+        return stub.downloadTable(downloadRequest);
+    }
+
+    public UploadResponse uploadTable(Table data) {
```
Comment (jason-dai) [inline scala/ppml/src/main/java/com/intel/analytics/bigdl/ppml/vfl/FGBoostStub.java:45]: uploadLabel

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    logger.debug(s"Update tree leaves ${(System.currentTimeMillis() - st) / 1000f} s")
+    st = System.currentTimeMillis()
+    trees.enqueue(currTree)
+    // Evaluate tree and update residual and grads (g and h)
+    validateLast(tree.dataset)
+    logger.debug(s"Validate last ${(System.currentTimeMillis() - st) / 1000f} s")
+    true
+  }
+
+  def trainRegressionTree(dataSet: Array[Tensor[Float]], indices: Array[Array[Int]], totalRound: Int): Unit = {
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostEstimator.scala:119]: line is too long; need to enable style check for PPML in github action

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostAggregator.scala [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+      } else {
+        currIndex = currIndex * 2 + 2
+      }
+    }
+    logger.debug("Reach leaf " + treeLeaves(currIndex))
+    treeLeaves(currIndex)
+  }
+
+
+  def predictWithTree(aggPredict: Array[Array[(String, Array[java.lang.Boolean])]]): Array[Float] = {
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/fgboost/VflGBoostAggregator.scala:125]: line is too long; need to enable style check for PPML in github action

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/example/vfl_gboost_regression/VflGBoostRegression.scala [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    // load args and get data
+    val dataPath = argv.dataPath
+    val rowKeyName = argv.rowKeyName
+    val learningRate = argv.learningRate
+
+
+    /**
+     * Usage of BigDL PPML starts from here
+     */
+    FLContext.initFLContext()
```
Comment (jason-dai) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/example/vfl_gboost_regression/VflGBoostRegression.scala:47]: `initFLContext` should call `initNNContext` internally

**Pull request merged**